syntax = "proto3";
package slog;

import "proto/machine_id.proto";

enum TransactionType {
    UNKNOWN = 0;
    SINGLE_HOME = 1;
    MULTI_HOME = 2;
    LOCK_ONLY = 3;
}

enum TransactionStatus {
    NOT_STARTED = 0;
    COMMITTED = 1;
    ABORTED = 2;
    ABORTED_WITHOUT_LOCKS = 3; // Remaster caused abort
}

message MasterMetadata {
    uint32 master = 1;
    uint32 counter = 2;
}

message TransactionInternal {
    uint32 id = 1;
    map<string, MasterMetadata> master_metadata = 2;
    TransactionType type = 3;
    internal.MachineId coordinating_server = 4;
}

message Transaction {
    TransactionInternal internal = 1;
    
    map<string, string> read_set = 2;
    map<string, string> write_set = 3;

    TransactionStatus status = 4;
    string abort_reason = 5;
    repeated string delete_set = 6;

    oneof procedure {
        /*
        Commands:
        GET key1
        SET key2 value2
        DEL key4
        COPY key1 key3
        */
        string code = 7;

        /*
        Remaster transactions will move the keys in the write_set to the new master.
        MasterMetadata must still be correct for the keys.
        */
        uint32 new_master = 8;
    }
}